TSM: Handle concurrent deletes for compaction

joelegasse opened the pull request:
This should handle concurrent calls to `DeleteSeriesRange` and ensure that compactions aren't restarted until all requests have been handled.

This may also help with #7168, but @e-dard would have to give this a look to see.

jwilder commented on tsdb/engine/tsm1/engine.go:
```diff
@@ -154,80 +154,87 @@ func (e *Engine) SetEnabled(enabled bool) {
 }
 
-func (e *Engine) disableLevelCompactions() {
+func (e *Engine) disableLevelCompactions(n int) {
```
What is `n`?  I see `0` and `1` passed in as values, but I don't understand what it means.

joelegasse replied:
`n` represents the number of workers that are expecting compactions to remain disabled. Since this can be called externally from `SetCompactionsEnabled`, and we can't guarantee the same call pattern as we can when used internally, I went with not tracking the status of external calls to enable/disable.

Thinking about it some more, we could try to track the state that the external calls think the status should be, and use that as a "worker"

jwilder commented on tsdb/engine/tsm1/engine.go:
```diff
@@ -57,13 +57,13 @@ const (
+	mu           sync.RWMutex
+	wg           sync.WaitGroup
+	done         chan struct{}
+	levelWorkers int
```
Is this the number of running goroutines?  Can you add a comment about this field?

joelegasse replied:
I'll add a comment, but it's meant to represent the number of working goroutines that expect compactions to be in a disabled state.

joelegasse commented:
@jwilder I updated some comments, and switched from an `int` parameter to a simple `bool` flag indicating whether or not the calls should wait (or correspond to) the opposite call.

Please let me know if the wording doesn't make sense, or you'd rather see the comments elsewhere. I don't want future us to be pissed at present me... :-)

jwilder commented:
👍
